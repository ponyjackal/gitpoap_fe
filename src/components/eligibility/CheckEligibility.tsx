import React, { useState } from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Box, BoxProps, Stack } from '@mantine/core';
import { Button, Header, Input, Loader, Text } from '../shared/elements';
import { EligibleClaimsQuery, useEligibleClaimsQuery } from '../../graphql/generated-gql';
import { FaEthereum, FaSearch } from 'react-icons/fa';
import { ClaimItem } from './ClaimItem';
import { useRouter } from 'next/router';
import { useWeb3Context } from '../wallet/Web3Context';
import { useUrlState } from '../../hooks/useUrlState';

export const ClaimListContainer = styled(Box)<BoxProps>`
  display: grid;
  width: 100%;
  grid-template-columns: repeat(3, 1fr);
  column-gap: ${rem(16)};
  row-gap: ${rem(32)};

  @media (max-width: ${rem(1600)}) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media (max-width: ${rem(1000)}) {
    grid-template-columns: repeat(1, 1fr);
  }
`;

export type Claim = Exclude<EligibleClaimsQuery['claims'], undefined | null>[number];

type QueryVars = {
  page: number;
  perPage: number;
};

export const CheckEligibility = () => {
  const router = useRouter();
  const urlSearchQuery = router.query.search as string | undefined;
  const { connectionStatus, connect } = useWeb3Context();
  const { value, setValue, debouncedValue } = useUrlState('search');
  const [variables] = useState<QueryVars>({
    page: 1,
    perPage: 36,
  });
  const [result] = useEligibleClaimsQuery({
    variables: {
      query: decodeURIComponent(urlSearchQuery ?? ''),
      take: variables.perPage,
      skip: (variables.page - 1) * variables.perPage,
    },
  });
  const allClaims = result.data?.claims;

  if (result.error) {
    return null;
  }

  return (
    <Stack align="center" mt={rem(80)} justify="center">
      <Stack align="center">
        <Header>{'Check Eligibility'}</Header>
        <Text align="center" size="md">
          {"Search for your GitHub handle & check if you're eligible for any GitPOAPs"}
        </Text>
        {connectionStatus === 'disconnected' && (
          <Button leftIcon={<FaEthereum size={16} />} onClick={() => connect()}>
            {'Sign In'}
          </Button>
        )}
      </Stack>
      <Input
        mb={rem(40)}
        px={rem(15)}
        style={{ width: rem(500) }}
        placeholder={'VBUTERIN...'}
        value={value}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => setValue(e.target.value)}
        icon={result.fetching && !!debouncedValue ? <Loader size={18} /> : <FaSearch />}
      />
      {allClaims &&
      allClaims.length === 0 &&
      !result.fetching &&
      debouncedValue &&
      debouncedValue.length > 0 ? (
        <Text style={{ marginBottom: rem(40), fontSize: rem(18) }}>
          {'No unminted GitPOAPs found'}
        </Text>
      ) : (
        <ClaimListContainer mt={rem(50)} mb={rem(55)}>
          <>
            {allClaims &&
              debouncedValue &&
              result.operation?.variables.query &&
              result.operation?.variables.query.length > 0 &&
              allClaims.map((claim, i) => {
                return <ClaimItem key={i} claim={claim} />;
              })}
          </>
        </ClaimListContainer>
      )}
    </Stack>
  );
};
