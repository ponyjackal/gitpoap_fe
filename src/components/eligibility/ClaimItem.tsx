import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { EligibleClaimsQuery } from '../../graphql/generated-gql';
import { Group, Stack, Text } from '@mantine/core';
import { GitPOAP } from '../shared/compounds/GitPOAP';
import { Header } from '../shared/elements';
import { truncateString } from '../../helpers';
import { Link } from '../shared/compounds/Link';

const Container = styled(Group)`
  max-width: 100%;
`;

const Title = styled(Header)`
  text-align: left;
  font-size: ${rem(24)};
  line-height: ${rem(28)};
`;

export type Claim = Exclude<EligibleClaimsQuery['claims'], undefined | null>[number];

type ClaimItemProps = {
  claim: Claim;
};

const getIssuedTo = ({ issuedAddress, email, user }: Claim) => {
  if (issuedAddress?.ensName) {
    return issuedAddress.ensName;
  } else if (issuedAddress?.ethAddress) {
    return issuedAddress.ethAddress;
  } else if (email) {
    return email.emailAddress;
  } else if (user?.githubHandle) {
    return user.githubHandle;
  }

  return '';
};

export const ClaimItem = ({ claim }: ClaimItemProps) => {
  const issuedTo = getIssuedTo(claim);
  const hasRepo = claim.gitPOAP.project && claim.gitPOAP.project.repos.length > 0;
  const fullRepoName = hasRepo
    ? `${claim.gitPOAP.project?.repos[0].organization.name}/${claim.gitPOAP.project?.repos[0].name}`
    : '';

  return (
    <Container position="center" align="start" noWrap p={rem(10)}>
      <Stack align="start">
        <GitPOAP gitPOAPId={claim.gitPOAP.id} imgSrc={claim.gitPOAP.imageUrl} size="sm" />
      </Stack>
      <Stack align="flex-start" justify="flex-start" spacing={10}>
        <Title>{claim.gitPOAP.name.replace('GitPOAP: ', '')}</Title>

        <Text align="left">{claim.gitPOAP.description}</Text>
        <Stack spacing={0}>
          <Group position="left" align="flex-start" spacing="xs">
            <Text weight={600}>{'Issued to: '}</Text>
            <Text>{truncateString(issuedTo, 24)}</Text>
          </Group>
          {hasRepo && (
            <Group spacing="xs">
              <Text weight={600}>{'For: '}</Text>
              <Link href={`/gh/${fullRepoName}`}>
                <Text weight="bold" variant="link">{`${fullRepoName}`}</Text>
              </Link>
            </Group>
          )}
        </Stack>
      </Stack>
    </Container>
  );
};
