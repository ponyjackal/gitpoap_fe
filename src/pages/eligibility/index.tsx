import React from 'react';
import { useRouter } from 'next/router';
import Head from 'next/head';
import { Grid } from '@mantine/core';
import { NextPageWithLayout } from '../_app';
import { CheckEligibility } from '../../components/eligibility/CheckEligibility';

const CheckEligibilityPage: NextPageWithLayout = () => {
  const router = useRouter();
  const searchQuery = router.query.search as string | undefined;

  return (
    <>
      <Head>
        <title>{`${
          searchQuery ? `${searchQuery} - Check Eligibility` : 'Check Eligibility'
        } | GitPOAP`}</title>
        <meta
          name="GitPOAP is a decentralized reputation platform that represents off-chain accomplishments and contributions on chain as POAPs."
          content="GitPOAP Eligibility Check"
        />
      </Head>
      <Grid justify="center">
        <Grid.Col xs={11} sm={11} md={11} lg={10} xl={10}>
          <CheckEligibility />
        </Grid.Col>
      </Grid>
    </>
  );
};

export default CheckEligibilityPage;
