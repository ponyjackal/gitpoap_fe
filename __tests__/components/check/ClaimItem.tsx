import { render } from '@testing-library/react';
import 'jest-styled-components';
import { ClaimItem, Claim } from '../../../src/components/eligibility/ClaimItem';

const CLAIM: Claim = {
  id: 1,
  issuedAddress: null,
  githubUser: {
    githubHandle: 'peebeejay',
  },
  gitPOAP: {
    id: 1,
    name: 'GitPOAP: Test Name of the GitPOAP',
    description: 'Test description',
    imageUrl: 'https://test.com',
    project: {
      repos: [
        {
          name: 'test-repo',
          organization: {
            name: 'test-org',
          },
        },
      ],
    },
  },
};

describe('ClaimItem', () => {
  it('should render correctly', () => {
    const { container } = render(<ClaimItem claim={CLAIM} />);
    expect(container).toMatchSnapshot();

    const item = container.firstChild;
    expect(item).toBeInTheDocument();
  });

  it('should render the correct text', () => {
    const { getByText } = render(<ClaimItem claim={CLAIM} />);
    expect(getByText('peebeejay')).toBeInTheDocument();
    expect(getByText('Test Name of the GitPOAP')).toBeInTheDocument();
    expect(getByText('Test description')).toBeInTheDocument();
  });

  /* There is the eventuality that the claim doesn't have an associated project / repo / org. */
  it('should render the org and repo name if specified', () => {
    const { getByText } = render(<ClaimItem claim={CLAIM} />);
    expect(getByText('test-org/test-repo')).toBeInTheDocument();

    expect(getByText('test-org/test-repo').closest('a')).toHaveAttribute(
      'href',
      '/gh/test-org/test-repo',
    );
  });
});
