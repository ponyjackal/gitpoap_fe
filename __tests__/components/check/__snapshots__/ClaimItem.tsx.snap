// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClaimItem should render correctly 1`] = `
.c5 {
  -webkit-transition: 150ms background-color ease-in-out,150ms opacity ease-in-out;
  transition: 150ms background-color ease-in-out,150ms opacity ease-in-out;
  -webkit-clip-path: url('#hexagonPath');
  clip-path: url('#hexagonPath');
  overflow: hidden;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}

.c3 {
  -webkit-transition: 150ms background-color ease-in-out,150ms opacity ease-in-out;
  transition: 150ms background-color ease-in-out,150ms opacity ease-in-out;
  -webkit-clip-path: url('#hexagonPath');
  clip-path: url('#hexagonPath');
  overflow: hidden;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}

.c9 {
  --s: 9.375rem;
  position: absolute;
  top: 0.1875rem;
  left: 0.1875rem;
  width: var(--s);
  height: calc(var(--s) * 1);
  display: inline-block;
  font-size: initial;
  background: #292B3D;
}

.c6 {
  position: relative;
  --s: 10.125rem;
  width: var(--s);
  height: calc(var(--s) * 1);
  background: #E2E2EE;
}

.c6:hover:not([disabled]) {
  cursor: pointer;
  background-color: #5596F6;
}

.c6:hover:not([disabled]) .c8 {
  opacity: 0.7;
}

.c6:active:not([disabled]) {
  background-color: #1761CF;
}

.c6:active:not([disabled]) .c8 {
  opacity: 0.5;
}

.c6[disabled] {
  cursor: not-allowed;
  background-color: #838495;
}

.c6[disabled] .c8:before {
  background: #10111E;
}

.c7 {
  --s: 9.75rem;
  position: absolute;
  top: 0.1875rem;
  left: 0.1875rem;
  width: var(--s);
  height: calc(var(--s) * 1);
  background: #10111E;
}

.c4 {
  --s: 10.125rem;
  width: var(--s);
  height: calc(var(--s) * 1);
}

.c11 {
  display: inline-block;
  font-weight: 700;
  font-size: 0.8125rem;
  line-height: 1.0625rem;
  -webkit-letter-spacing: 0.03125rem;
  -moz-letter-spacing: 0.03125rem;
  -ms-letter-spacing: 0.03125rem;
  letter-spacing: 0.03125rem;
  text-align: center;
  color: #307AE8;
  -webkit-transition: color 150ms ease-in-out;
  transition: color 150ms ease-in-out;
  cursor: pointer;
  color: #307AE8;
}

.c12:hover:not([disabled]) {
  color: #5596F6;
}

.c12:active:not([disabled]) {
  color: #1761CF;
}

.c12[disabled] {
  color: #393A51;
  cursor: not-allowed;
}

.c15 {
  font-family: VT323;
  font-size: 3rem;
  line-height: 3rem;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 9.375rem;
}

.c13 {
  margin-top: 0.625rem;
  text-align: center;
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
}

.c14 {
  font-family: PT Mono;
  font-style: normal;
  font-weight: bold;
  font-size: 0.6875rem;
  line-height: 1.125rem;
  text-align: center;
  -webkit-letter-spacing: 0.075rem;
  -moz-letter-spacing: 0.075rem;
  -ms-letter-spacing: 0.075rem;
  letter-spacing: 0.075rem;
  text-transform: uppercase;
  color: #E2E2EE;
  margin-top: 0.5rem;
  -webkit-transition: color 200ms ease-in-out;
  transition: color 200ms ease-in-out;
}

.c2 {
  position: relative;
}

.c0 {
  max-width: 100%;
}

.c16 {
  text-align: left;
  font-size: 1.5rem;
  line-height: 1.75rem;
}

@media (max-width:56.25rem) {
  .c15 {
    font-size: 2.5rem;
  }
}

.emotion-1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  gap: 16px;
  padding: 0.625rem;
}

.emotion-1>* {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: top;
  -ms-flex-pack: top;
  -webkit-justify-content: top;
  justify-content: top;
  gap: 16px;
}

.emotion-5 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  color: inherit;
  font-size: inherit;
  line-height: 1.55;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-5:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-5:focus:not(:focus-visible) {
  outline: none;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  gap: 10px;
}

.emotion-11 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  color: inherit;
  font-size: inherit;
  line-height: 1.55;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: left;
}

.emotion-11:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-11:focus:not(:focus-visible) {
  outline: none;
}

.emotion-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: top;
  -ms-flex-pack: top;
  -webkit-justify-content: top;
  justify-content: top;
  gap: 0;
}

.emotion-15 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  gap: 10px;
}

.emotion-15>* {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
}

.emotion-17 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  color: inherit;
  font-size: inherit;
  line-height: 1.55;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 600;
}

.emotion-17:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-17:focus:not(:focus-visible) {
  outline: none;
}

.emotion-21 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  gap: 10px;
}

.emotion-21>* {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
}

.emotion-25 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  color: #1c7ed6;
  font-size: inherit;
  line-height: 1.55;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: bold;
}

.emotion-25:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-25:focus:not(:focus-visible) {
  outline: none;
}

@media (hover: hover) {
  .emotion-25:hover {
    -webkit-text-decoration: underline;
    text-decoration: underline;
  }
}

@media (hover: none) {
  .emotion-25:active {
    -webkit-text-decoration: underline;
    text-decoration: underline;
  }
}

<div>
  <div
    class="emotion-0 c0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="c1"
      >
        <div
          class="c1 c2"
        >
          <a
            class="c3 c4"
            href="/gp/1"
          >
            <div
              class="c5 c6"
            >
              <div
                class="c5 c7"
              >
                <div
                  class="c8 c5 c9"
                >
                  <span
                    style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
                  >
                    <img
                      alt=""
                      data-nimg="fill"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                    />
                    <noscript />
                  </span>
                </div>
              </div>
            </div>
            <svg
              height="0"
              style="position: absolute;"
              width="0"
              xmlns="http://www.w3.org/2000/svg"
            >
              <clippath
                clipPathUnits="objectBoundingBox"
                id="hexagonPath"
              >
                <path
                  d="M 0.3857 0.038 Q 0.5 -0.028 0.6143 0.038 L 0.8429 0.17 Q 0.9573 0.236 0.9573 0.368 L 0.9573 0.632 Q 0.9573 0.764 0.8429 0.83 L 0.6143 0.962 Q 0.5 1.028 0.3857 0.962 L 0.1571 0.83 Q 0.0427 0.764 0.0427 0.632 L 0.0427 0.368 Q 0.0427 0.236 0.1571 0.17 Z"
                />
              </clippath>
            </svg>
          </a>
        </div>
        <div
          class="c10"
        >
          <a
            href="/gp/1"
          >
            <div
              class="emotion-4 c11 c12 c13 emotion-5"
            />
          </a>
          <div
            class="c14"
          />
        </div>
      </div>
    </div>
    <div
      class="emotion-2 emotion-7"
    >
      <div
        class="emotion-4 c15 c16 emotion-5"
      >
        Test Name of the GitPOAP
      </div>
      <div
        class="emotion-4 emotion-11"
      >
        Test description
      </div>
      <div
        class="emotion-2 emotion-13"
      >
        <div
          class="emotion-0 emotion-15"
        >
          <div
            class="emotion-4 emotion-17"
          >
            Issued to: 
          </div>
          <div
            class="emotion-4 emotion-5"
          >
            peebeejay
          </div>
        </div>
        <div
          class="emotion-0 emotion-21"
        >
          <div
            class="emotion-4 emotion-17"
          >
            For: 
          </div>
          <a
            href="/gh/test-org/test-repo"
          >
            <div
              class="emotion-4 emotion-25"
            >
              test-org/test-repo
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
